flutter/lib/src/span_frame_metrics_collector.dart

```diff
@@ -0,0 +1,253 @@
+    frames[options.clock()] = frameDuration;
+
+    _stopwatch.reset();
+
+    if (_frameCallbackHandler.hasScheduledFrame == true) {
+      _stopwatch.start();
+    }
+  }
+
+  void _applyFrameMetricsToSpan(
+      ISentrySpan span, Map<String, int> frameMetrics) {
+    frameMetrics.forEach((key, value) {
+      span.setData(key, value);
+    });
+
+    // This will call the methods on the tracer, not on the span directly
+    if (span is SentrySpan && span.isRootSpan) {
+      frameMetrics.forEach((key, value) {
+        // ignore: invalid_use_of_internal_member
+        span.tracer.setData(key, value);
+
+        // In measurements we change e.g frames.total to frames_total
+        // We don't do span.tracer.setMeasurement because setMeasurement in SentrySpan
+        // uses the tracer internally
+        span.setMeasurement(key.replaceAll('.', '_'), value);
+      });
+    }
+  }
+
+  @visibleForTesting
+  Map<String, int> calculateFrameMetrics(
+      ISentrySpan span, DateTime spanEndTimestamp, int displayRefreshRate) {
+    if (frames.isEmpty) {
+      options.logger(
+          SentryLevel.info, 'No frame durations available in frame tracker.');
+      return {};
+    }
+
+    final expectedFrameDuration = ((1 / displayRefreshRate) * 1000).toInt();
+
+    int slowFramesCount = 0;
+    int frozenFramesCount = 0;
+    int slowFramesDuration = 0;
+    int frozenFramesDuration = 0;
+    int framesDelay = 0;
+
+    for (final entry in frames.entries) {
+      final frameDuration = entry.value;
+      final frameEndTimestamp = entry.key;
+      final frameStartMs =
+          frameEndTimestamp.millisecondsSinceEpoch - frameDuration;
+      final frameEndMs = frameEndTimestamp.millisecondsSinceEpoch;
+      final spanStartMs = span.startTimestamp.millisecondsSinceEpoch;
+      final spanEndMs = spanEndTimestamp.millisecondsSinceEpoch;
+
+      final frameFullyContainedInSpan =
+          frameEndMs <= spanEndMs && frameStartMs >= spanStartMs;
+      final frameStartsBeforeSpan =
+          frameStartMs < spanStartMs && frameEndMs > spanStartMs;
+      final frameEndsAfterSpan =
+          frameStartMs < spanEndMs && frameEndMs > spanEndMs;
+      final framePartiallyContainedInSpan =
+          frameStartsBeforeSpan || frameEndsAfterSpan;
+
+      int effectiveDuration = 0;
+      int effectiveDelay = 0;
+
+      if (frameFullyContainedInSpan) {
+        effectiveDuration = frameDuration;
+        effectiveDelay = max(0, frameDuration - expectedFrameDuration);
+      } else if (framePartiallyContainedInSpan) {
+        final intersectionStart = max(frameStartMs, spanStartMs);
+        final intersectionEnd = min(frameEndMs, spanEndMs);
+        effectiveDuration = intersectionEnd - intersectionStart;
+
+        final fullFrameDelay = max(0, frameDuration - expectedFrameDuration);
+        final intersectionRatio = effectiveDuration / frameDuration;
+        effectiveDelay = (fullFrameDelay * intersectionRatio).round();
+      } else {
```

Comment:
So we don't check all frames
```suggestion
      } else if (frameStartMs > spanEndMs) {
        // Other frames will be newer than this span, as frames are ordered
        break;
      } else {
```